### PR TITLE
Fix issue with devirtualization and tailcalls

### PR DESCRIPTION
**Pull request by briansull**
When performing devirtualization we can not do both an unboxing optimization and a tail call optimization
-  Explicit tail calls are now checked for and blocked from performing an unboxing operation in impDevirtualizeCall.
-  If late devirtualization calls impDevirtualizeCall with an IMPLICIT_TAILCALL we will clear this flag if we decide to perform the unboxing operation.

### CODE REVIEW

**Comment by briansull** [comment]
@AndyAyersMS FYI

**Comment by AndyAyersMS** [comment]
You should make it clearer that this restriction is for explicit tail calls only, where presumably the IL producer is confident that at the point of the explicit tail call there aren't any live references to the caller frame, and if we undo the box, we will introduce one.

And I think this could also happen for late devirt though it would require a different kind of test case.

Seems like you could just check for `GTF_CALL_M_EXPLICIT_TAILCALL` to inhibit the box undo transform, instead of passing in extra info, and handle it more cleanly and more generally.

**Comment by briansull** [comment]
The flag GTF_CALL_M_EXPLICIT_TAILCALL is current not yet set by the importer when we reach this method.
It currently gets set here: importer.cpp line 8753 and the call to devirtualize is on line 8597.

**Comment by briansull** [comment]
The IL with  for the method is:

```
****** START compiling Program:Test(int):ref:this (MethodHash=f85425b1)
IL to import:
IL_0000  03                ldarg.1     
IL_0001  8c 04 00 00 01    box          0x1000004
IL_0006  fe 14             tail.       
IL_0008  6f 03 00 00 0a    callvirt     0xA000003
IL_000d  2a                ret         
```

**Comment by AndyAyersMS** [comment]
Hmm, ok. We should still fix the late devirt path, and that one needs to work with the call flags. So we need to adapt for one convention or the other.

If it was me I'd probably handle this by passing in a bool `isExplicitTailCall` and returning a bool result indicating if we unboxed, and then adapt the two call sites to suit.

**Comment by briansull** [comment]
@AndyAyersMS @jashook PTAL

File: src/jit/importer.cpp [hunk 1]
```diff
@@ -8742,7 +8744,6 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
 
             if (info.compCompHnd->canTailCall(info.compMethodHnd, methHnd, exactCalleeHnd, explicitTailCall))
             {
-                canTailCall = true;
```
Comment (briansull) [inline src/jit/importer.cpp]: This assignment was redundant (see line 8736 above)

**Comment by briansull** [comment]
@dotnet-bot Test Ubuntu x64 Checked Innerloop Build and Test

**Comment by briansull** [comment]
@dotnet-bot test Ubuntu x64 Checked Innerloop Build and Test (Jit - TieredCompilation=0)

File: src/jit/importer.cpp [hunk 2]
```diff
@@ -20558,6 +20561,12 @@ void Compiler::impDevirtualizeCall(GenTreeCall*            call,
     {
         JITDUMP("Now have direct call to boxed entry point, looking for unboxed entry point\n");
 
+        if (isTailCall)
+        {
+            JITDUMP("Call is an explcit tail call, we cannot perform an unbox\n");
```
Comment (AndyAyersMS) [inline src/jit/importer.cpp]: Typo: `explcit`

File: src/jit/importer.cpp [hunk 3]
```diff
@@ -20174,6 +20175,7 @@ bool Compiler::IsMathIntrinsic(GenTree* tree)
 //     contextHandle -- [IN/OUT] context handle for the call. Updated iff call devirtualized.
 //     exactContextHnd -- [OUT] updated context handle iff call devirtualized
 //     isLateDevirtualization -- if devirtualization is happening after importation
+//     isTailCall -- [IN/OUT] true if we plan on using a tail call
```
Comment (AndyAyersMS) [inline src/jit/importer.cpp]: This is no longer [in/out] -- also rename to `isExplictTailCall`

**Comment by briansull** [comment]
Yes, I do have an IL test case that I will add

File: src/jit/compiler.h [hunk 4]
```diff
@@ -3329,7 +3329,8 @@ class Compiler
                              unsigned*               methodFlags,
                              CORINFO_CONTEXT_HANDLE* contextHandle,
                              CORINFO_CONTEXT_HANDLE* exactContextHandle,
-                             bool                    isLateDevirtualization);
+                             bool                    isLateDevirtualization,
+                             bool                    isTailCall);
```
[the remark this answers is not in the document]
Comment (briansull) [reply src/jit/compiler.h]: This parameter is used and it is also renamed as isExplicitTailCall

**Comment by briansull** [comment]
@dotnet-bot Test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by briansull** [comment]
@dotnet-bot Test Windows_NT arm Cross Checked Innerloop Build and Test

**Pull request merged**